lib/kubernetes-deploy/container_logs.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+module KubernetesDeploy
+  class ContainerLogs
+    attr_reader :lines, :container_name
+
+    DEFAULT_LINE_LIMIT = 250
+
+    def initialize(parent_id:, container_name:, logger:)
+      @parent_id = parent_id
+      @container_name = container_name
+      @logger = logger
+      @lines = []
+      @next_print_index = 0
+    end
+
+    def sync(kubectl)
+      new_logs = fetch_latest(kubectl)
+      return unless new_logs.present?
+      @lines += deduplicate(new_logs)
+    end
+
+    def empty?
+      lines.empty?
+    end
+
+    def print_latest(prefix: false)
+      prefix_str = "[#{container_name}]  " if prefix
+
+      lines[@next_print_index..-1].each do |msg|
+        @logger.info "#{prefix_str}#{msg}"
+      end
+
+      @next_print_index = lines.length
+    end
+
+    def print_all
+      lines.each { |line| @logger.info("\t#{line}") }
+    end
+
+    private
+
+    def fetch_latest(kubectl)
+      cmd = ["logs", @parent_id, "--container=#{container_name}", "--timestamps"]
+      cmd << if @last_timestamp.present?
+        "--since-time=#{rfc3339_timestamp(@last_timestamp)}"
+      else
+        "--tail=#{DEFAULT_LINE_LIMIT}"
+      end
+      out, _err, _st = kubectl.run(*cmd, log_failure: false)
+      out.split("\n")
+    end
+
+    def rfc3339_timestamp(time)
+      time.strftime("%FT%T.%N%:z")
+    end
+
+    def deduplicate(logs)
+      deduped = []
+      timestamps = []
+      check_for_duplicate = true
+
+      logs.each do |line|
+        timestamp, msg = split_timestamped_line(line)
+        next if check_for_duplicate && likely_duplicate?(timestamp)
+        check_for_duplicate = false # logs are ordered, so once we've seen a new one, assume all subsequent logs are new
+        timestamps << timestamp if timestamp
```

Comment:
nit: I think you could just track the lastest timestamp, instead of all of them

Reply:
I just spent a couple hours debugging a weird sporadic failure on my other branch in the integration test checking for duplicates. Digging in I noticed that it was often (but not always) the last lines that were duplicated, and often (but not always) exactly 250 duplicates. Turns out that if we saw an entire batch of duplicates, the corresponding `timestamps.last` line would set `@last_timestamp` to `nil`, foiling the deduplication system. The simplest fix was to set the timestamp directly here instead of collecting them, as you suggested. (Note re: [this comment](https://github.com/Shopify/kubernetes-deploy/pull/340#discussion_r223246008):  the `check_for_duplicate` optimization makes it safe to put it back inside the loop)

Reply:
I put the fix and regression test in the other PR btw, since that one needed it to pass CI: https://github.com/Shopify/kubernetes-deploy/pull/337/commits/6d971000b91d3cf677ad0bca5ffa8b517ca266e2. I'll merge both back-to-back tomorrow.